packages/flutter/lib/src/services/platform_views.dart

```diff
@@ -427,35 +447,189 @@ enum _AndroidViewState {
+      toolMinor: event.radiusMinor,
+      touchMajor: event.radiusMajor,
+      touchMinor: event.radiusMinor,
+      x: position.dx,
+      y: position.dy,
+    );
+  }
+
+  void handlePointerUpEvent(PointerUpEvent event) {
+    pointerPositions.remove(event.pointer);
+    pointerProperties.remove(event.pointer);
+    if (pointerProperties.isEmpty) {
+      nextPointerId = 0;
+      downTimeMillis = null;
+    }
+  }
+
+  void handlePointerCancelEvent(PointerCancelEvent event) {
+    pointerPositions.clear();
+    pointerProperties.clear();
+    nextPointerId = 0;
+    downTimeMillis = null;
+  }
+
+  AndroidMotionEvent toAndroidMotionEvent(PointerEvent event) {
+    final List<int> pointers = pointerPositions.keys.toList();
+    final int pointerIdx = pointers.indexOf(event.pointer);
+    final int numPointers = pointers.length;
+
+    // This value must match the value in engine's FlutterView.java.
+    // This flag indicates whether the original Android pointer events were batched together.
+    const int kPointerDataFlagBatched = 1;
+
+    // Android MotionEvent objects can batch information on multiple pointers.
+    // Flutter breaks these such batched events into multiple PointerEvent objects.
+    // When there are multiple active pointers we accumulate the information for all pointers
+    // as we get PointerEvents, and only send it to the embedded Android view when
+    // we see the last pointer. This way we achieve the same batching as Android.
+    if (event.platformData == kPointerDataFlagBatched ||
+        (isSinglePointerAction(event) && pointerIdx < numPointers - 1)) {
+      return null;
+    }
+
+    int action;
+    switch (event.runtimeType) {
+      case PointerDownEvent:
+        action = numPointers == 1
+            ? AndroidViewController.kActionDown
+            : AndroidViewController.pointerAction(
+                pointerIdx, AndroidViewController.kActionPointerDown);
+        break;
+      case PointerUpEvent:
+        action = numPointers == 1
+            ? AndroidViewController.kActionUp
+            : AndroidViewController.pointerAction(
+                pointerIdx, AndroidViewController.kActionPointerUp);
+        break;
+      case PointerMoveEvent:
+        action = AndroidViewController.kActionMove;
+        break;
+      case PointerCancelEvent:
+        action = AndroidViewController.kActionCancel;
+        break;
+      default:
+        return null;
+    }
+
+    return AndroidMotionEvent(
+      downTime: downTimeMillis,
+      eventTime: event.timeStamp.inMilliseconds,
+      action: action,
+      pointerCount: pointerPositions.length,
+      pointerProperties: pointers
+          .map<AndroidPointerProperties>((int i) => pointerProperties[i])
+          .toList(),
+      pointerCoords: pointers
+          .map<AndroidPointerCoords>((int i) => pointerPositions[i])
+          .toList(),
+      metaState: 0,
+      buttonState: 0,
+      xPrecision: 1.0,
+      yPrecision: 1.0,
+      deviceId: 0,
```

Comment:
this should be `event.device` + unit test

Reply:
Actually, this isn't `event.device`. We will need to add a `deviceId` to `MotionEvent`, and most likely each of the Android-specific values. 

@goderbauer do you know if there's a recommended pattern for adding platform-specific data to `MotionEvent`? It looks like we will pretty much need to pass each of these constants.

None of these values are sent from https://github.com/flutter/engine/blob/712f619737f1323fd2908b397aa8bd64fb2ccbe5/shell/platform/android/io/flutter/embedding/android/AndroidTouchProcessor.java#L171

Reply:
Can we use a similar approach that @gspencergoog used to add platform-specific stuff to the keyboard events? See `RawKeyEventData` [1] and its platform-specific subclasses.

[1] https://master-api.flutter.dev/flutter/services/RawKeyEventData-class.html

Reply:
Well, I wouldn't go that route.  That's all there because of legacy that we didn't want to break. We're planning on removing it and moving it all to the platform side now.

Better would be to pass that data you need in a platform independent form and add it for all platforms.

Context: https://flutter.dev/go/platform-based-key-events

Reply:
It looks like `RawKeyEventData` is constructed from a system channel message, and the params are passed in a `Map`. 

Perhaps, the equivalent is to add a `Map` to `MotionEvent`, and then populate the `PointerDataPacket` in  the engine. https://github.com/flutter/engine/blob/637a8e7cd0440e5c97706aad89ba3f4decd432ee/lib/ui/hooks.dart#L294. 

Does this sound resonable?

Reply:
After studying this further, I agree with greg that adding platform-specific stuff to the pointer events like that is not a good idea.

Reply:
After some offline discussion, the conclusion was to add a unique identifier to `MotionEvent` that can be used to lookup the original motion event in the platform.